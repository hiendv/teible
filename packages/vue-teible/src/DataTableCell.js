import { dotGet } from 'teible'

export default {
  functional: true,
  props: {
    item: {
      type: Object,
      required: true
    },
    column: {
      type: Object,
      required: true
    }
  },
  render (h, { props, data }) {
    if (props.column.field) {
      let value = dotGet(props.item, props.column.field)
      if (props.column.scopedSlots && typeof props.column.scopedSlots.default === 'function') {
        return h('td', data, props.column.scopedSlots.default({ value, item: props.item, column: props.column }))
      }

      if (typeof value !== 'string') {
        value = JSON.stringify(value)
      }
      return h('td', data, value)
    }

    if (props.column.scopedSlots && typeof props.column.scopedSlots.default === 'function') {
      return h('td', data, props.column.scopedSlots.default(props))
    }

    return h('td', data, props.column.children)
  }
}
