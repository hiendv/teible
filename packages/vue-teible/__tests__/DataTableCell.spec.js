import { mount } from '@vue/test-utils'
import DataTableCell from '../src/DataTableCell.js'

describe('DataTableCell', () => {
  it('renders correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: 'foo',
            two: 'bar'
          },
          column: {
            field: 'one'
          },
          index: 0
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders objects correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: {
              foo: 'bar'
            },
            two: 'bar'
          },
          column: {
            field: 'one'
          },
          index: 0
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders scopedSlots correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: 'foo',
            two: 'bar'
          },
          column: {
            scopedSlots: {
              default (props) {
                return JSON.stringify(props.item)
              }
            }
          },
          index: 0
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders scopedSlots with field correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: 'foo',
            two: 'bar'
          },
          column: {
            field: 'one',
            scopedSlots: {
              default ({ value }) {
                return value
              }
            }
          },
          index: 0
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders children correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: 'foo',
            two: 'bar'
          },
          column: {
            children: ['yo']
          },
          index: 0
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders scopedSlots with index correctly', () => {
    const wrapper = mount(DataTableCell, {
      context: {
        props: {
          item: {
            one: 'foo',
            two: 'bar'
          },
          column: {
            scopedSlots: {
              default ({ index }) {
                return index
              }
            }
          },
          index: 1337
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
