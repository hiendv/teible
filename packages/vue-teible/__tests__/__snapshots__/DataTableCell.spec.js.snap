// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableCell renders children correctly 1`] = `<td>yo</td>`;

exports[`DataTableCell renders correctly 1`] = `<td>foo</td>`;

exports[`DataTableCell renders objects correctly 1`] = `<td>{"foo":"bar"}</td>`;

exports[`DataTableCell renders scopedSlots correctly 1`] = `<td>{"one":"foo","two":"bar"}</td>`;

exports[`DataTableCell renders scopedSlots with field correctly 1`] = `<td>foo</td>`;

exports[`DataTableCell renders scopedSlots with index correctly 1`] = `<td>1337</td>`;
